fastai/text/data.py

```diff
@@ -13,44 +13,106 @@
-        if self.item is not None:    return 1
-        if (self.n-1)%self.bptt == 0 or self.drop_last: return ((self.n-1) // self.bptt) * self.bs
-        return ((self.n-1) // self.bptt + 1) * self.bs
+    "Create a dataloader with bptt slightly changing."
     
+    class CircularIndex():
+        #When the index exceeds the length of self.idx then it is wrap to start at the head 
+        #or the end if indexing is moving backwards. The shuffled is done in-place, 
+        def __init__(self, length:int, forward:bool): 
+            self.idx      = np.arange(length)
+            self.forward_ = forward
+        def __getitem__(self, i): 
+            idx = self.idx
+            return idx[ i%len(idx) if self.forward_ else len(idx)-1-i%len(idx) ]
+        def __len__(self) -> int: return len(self.idx)
+        def shuffle(self): np.random.shuffle(self.idx)
+        def forward(self, forward:bool=True): self.forward_ = forward
+
+    def __init__(self, dataset:LabelList, bs:int=32, bptt:int=70, backwards:bool=False, shuffle:bool=False):
+        self.dataset,self.bs,self.bptt,self.shuffle,self.backwards = dataset,bs,bptt,shuffle,backwards
+        self.totalToks = 0
+        for rag in dataset.x.items: self.totalToks += len(rag)
+        self.ite_len   = self.bs*int( math.ceil( self.totalToks/(self.bptt*self.bs) )) if self.item is None else 1
+        self.idx = None
+
+    def __len__(self): return self.ite_len
     def __getattr__(self,k:str)->Any: return getattr(self.dataset, k)
-    
+   
+    def allocate_buffers(self):     
+        "allocate the required worth-case batchbuffer"
+        #idx:     used to index into items so that shuffle is used - if idx has been shuffled 
+        #ei:      index of the first rag to be extract. Returs as index to the next rag to be extracted
+        #eo:      index to the first token to be extracted in the first rag. Returs pointing to the next to be extract in the last rag
+        #         when iterating backwards then ei becomes 1+ the last token to be extraced in the rag 
+        #overlap: overlap=1 between batches, because we only predict the next token        
+        self.idx   = LanguageModelPreLoader.CircularIndex(len(self.dataset.x.items), not self.backwards)
+        self.batch = np.zeros( (self.bs, self.bptt+1), dtype=np.int64)
+        self.x, self.y = self.batch[:,0:self.bptt], self.batch[:,1:self.bptt+1]      
+        self.ei    = np.zeros(self.bs, dtype=np.int)
+        self.eo    = np.zeros(self.bs, dtype=np.int)
+
     def on_epoch_begin(self, **kwargs):
-        self.idxs = np.random.permutation(len(self.dataset)) if self.shuffle else arange_of(self.dataset)
-        self.text_idx = np.concatenate([[0],self.lengths[self.idxs].cumsum()])
-        
+        if self.idx is None: self.allocate_buffers()
+        if self.shuffle:     self.idx.shuffle()
+        self.idx.forward(self.backwards is False) 
+
+        #set up ei and eo to index the data so batches has continuous rows  
```

Comment:
This whole part can be vectorized (see the code you're replacing).

Reply:
I have passed hours on the existing code. The i idea here i to avoid allocations by slicing like so row[ibuf:ibuf+n] = rag[eo:eo+n]. this statement also handle type conversion implicitly if you passe np.uin16. 

I normallý go to great lengths to avoid loops i python and R because they are so slow. In this case using a loop is faster than vectorising. I guess the reason is that in most case the loop i only executed once. 

Let me make som measurements whether that is true

Reply:
measurements using bptt=70 shows that 48% of the calls to fill_forward pass 1 time through the loop, about 25% pass 2 times and the rest pass through the loop 3-7 times to build a row in the batch. If you have many short sentences it will take more loop than if you have sentence with 70 tokens or more. 
I think this explain why this approach is faster then the vectorisation in this special case